### DISCUSSION TITLE
routes 配置id 不生效

### DISCUSSION THREAD

**Discussion by ohroy** [umijs/umi#10182] [Q&A]
需求非常简单，类似 https://github.com/umijs/umi/blob/master/packages/plugins/src/layout.ts
即，在某个自定义layout里读取routes信息，然后遍历当前layout的子路由，把他们展示在菜单上。
这里面有一句关键代码
https://github.com/umijs/umi/blob/2c360ade8133d0f97ef3739144504169e26ba1aa/packages/plugins/src/layout.ts#L212
用来判断`route.id` ， 显然route应该是有一个id。望文生义，在在config里配置如下
```
    {
        name: 'userLayout',
        id: 'userLayout',
        path: '/user',
        component: '@/layouts/user',
    }
```
结果，结果不生效，打印结果 route.id 为 `7`， 也就是说，它还是不认我配置的，继而自动生成了一个。
虽然这里可以修改为其他参数比如name判断，但是总归是没有id可靠。 不知是设计成不识别这里的id，还是有bug？

------

注： 若使用插件api动态添加route，则可以指定id

**Comment by ohroy** [comment]
另外，对于当前最新版本来说，上述代码也会引发类型错误，原因是 `useAccessMarkedRoutes` 所需求的参数类型是`IRoute`，而`useAppData`返回的`clientRoutes`的类型是`IClientRoute`

**Comment by ohroy** [comment]
另外，
https://github.com/umijs/umi/blob/2c360ade8133d0f97ef3739144504169e26ba1aa/packages/plugins/src/layout.ts#L217
[此处](https://github.com/umijs/umi/blob/2c360ade8133d0f97ef3739144504169e26ba1aa/packages/plugins/src/layout.ts#L217)获取当前route的实现应该也存在小小瑕疵，考虑以下情况


config.ts -> routes
```ts
 {
        name: 'userLayout',
        id: 'userLayout',
        path: '/user', // 注意此行
        component: '@/layouts/user',
        routes: [{
            path: 'home', //  注意此行
            name: 'userHome',
            component: 'user/home',
            test: 123
        }]
}
```
此时，父路由使用了 `/user`的绝对路径，子路由使用了相对路径`home`, 我们通过 `/user/home`可以成功的访问到子路由，但是，此时由于子路由的path是home而非 /user/home，若调用
```ts
matchRoutes(route.children, location.pathname)
```
则返回 `undefined`

此处可能考虑需要这种情况，在matchRoutes之前动态组合出完整路径，方便后续matchRoutes调用

**Comment by fz6m** [comment]
id 是内部使用的标识，不能自己覆盖定义，可以定义到其他属性上来判断。

`matchRoutes` 可以匹配到数据的，你说可能会 `undefined` 给一个最小复现看看。
